src/ray/util/ordered_set.h

```diff
@@ -0,0 +1,44 @@
+#ifndef RAY_UTIL_ORDERED_SET_H
+#define RAY_UTIL_ORDERED_SET_H
+
+#include <list>
+#include <unordered_map>
+
+template <typename T>
+class ordered_set {
```

Comment:
This class is subtly buggy; you need to suppress the default `copy-constructor` and `copy-assignment` for this class. (Any guesses why?)

Reply:
Hm, probably cause of the iterators.

Reply:
Yup. Copying is not a compositional operation. If `A` and `B` have correct copy semantics that doesn't imply an a composition `struct S { A a; B b; };` of them does.

Reply:
Why is that so? My code gets much more awkward if there is no copy/assignment constructor so I hope we can bring it back somehow.

Reply:
@pcmoritz because the contents of one component of the class might be referenced by the other, so copying them will make them still point to the old instances' data. It should be more plausible if you consider that copying is inherently a sophisticated operation (imagine trying to copy a potentially cyclic graph of objects) -- which is something that's not really pointed out by those teaching C++. In this case it's a DAG but the fact that it's not a tree is what makes it difficult.

Reply:
In this particular case the iterators would point to the elements in old list rather than the copy.

Reply:
You could implement copying manually. The default implementation just won't work.

Reply:
Ok cool yeah, I realized that shortly after I wrote it. I deleted the copy constructors of the classes that use it now too as a fix :)

Reply:
I think you technically wouldn't need to delete the copy constructors of the classes that use it since they wouldn't exist if they can't be generated automatically?